1-Draft/RFC-ArgumentCompleter-BaseClass.md

```diff
@@ -0,0 +1,348 @@
+        /// <param name="resultType">the type of completion result</param>
+        private void CompleteIfAnyStartsWithWordToComplete(string text, string? listItemText = null,
+            string? toolTip = null, CompletionResultType resultType = CompletionResultType.ParameterValue)
+        {
+            if (StartWithWordToComplete(text: text) || StartWithWordToComplete(text: listItemText) || StartWithWordToComplete(text: toolTip))
+                Complete(text: text, listItemText ?? text, toolTip ?? text,
+                    resultType: resultType);
+        }
+
+        /// <summary>
+        ///     Adds a completion result to the result set if the any string argument starts with <see cref="WordToComplete" />
+        /// </summary>
+        /// <para>The comparison is case insensitive</para>
+        /// <param name="text">the text to be used as the auto completion result</param>
+        /// <param name="listItemText">the text to be displayed in a list</param>
+        /// <param name="toolTip">the text for the tooltip with details to be displayed about the object</param>
+        /// <param name="resultType">the type of completion result</param>
+        private void CompleteIfAnyContainsWordToComplete(string text, string? listItemText = null,
+            string? toolTip = null, CompletionResultType resultType = CompletionResultType.ParameterValue)
+        {
+            if (ContainsWordToComplete(text: text) || ContainsWordToComplete(text: listItemText) || ContainsWordToComplete(text: toolTip))
+                Complete(text: text, listItemText ?? text, toolTip ?? text,
+                    resultType: resultType);
+        }
+
+        /// <summary>
+        ///     Adds a completion result to the result set if the text contains <see cref="WordToComplete" />
+        /// </summary>
+        /// <para>The comparison is case insensitive</para>
+        /// <param name="text">the text to be used as the auto completion result</param>
+        /// <param name="listItemText">the text to be displayed in a list</param>
+        /// <param name="toolTip">the text for the tooltip with details to be displayed about the object</param>
+        /// <param name="resultType">the type of completion result</param>
+        private void CompleteIfTextContainsWordToComplete(string text, string? listItemText = null,
+            string? toolTip = null, CompletionResultType resultType = CompletionResultType.ParameterValue)
+        {
+            if (ContainsWordToComplete(text: text))
+                Complete(text: text, listItemText ?? text, toolTip ?? text,
+                    resultType: resultType);
+        }
+
+        /// <summary>
+        /// If necessary, puts quotation marks around the completion text
+        /// </summary>
+        /// <param name="text">The text to complete</param>
+        /// <returns></returns>
+        protected virtual string QuoteCompletionText(string text) => text.Contains(" ") ? $@"""{text}""" : text;
```

Comment:
What if the user is using a single quote?

Reply:
Oh gosh, had some flashbacks of working on argument completion for PSWordCloud. There's a bit more you need to handle for this kind of thing as well. Let me find some code I wrote a while back... here:

https://github.com/vexx32/PSWordCloud/blob/main/Module/src/PSWordCloud/Completers/FontFamilyCompleter.cs

This one covers escaping things that need escaping, spaces, `#` characters in the completion results (I happened to end up having fonts with that character on my computer and it breaks completions terribly 😀) and probably a few other things. I haven't managed to break it yet.

The only thing I don't think it bothers to handle is whether the user used single quotes, it currently replaces them with double quotes instead I think. So it could be expanded on to detect that and act accordingly (less things need escaping there, etc).

Reply:
I expect we have to turn this into a much more complete implementation. Maybe we can make use of the parser to determine what needs to be escaped?